[WIP] Allow nested categories with recursive navigation.

galli-leo opened the pull request:
## Overview
So I started working on allowing nested categories for custom categories. Currently there is no limit on the amount of nesting (i.e. levels). This is achieved through nested partials as well as keeping track of the level, and using that in the header tag (`<h{{level}}>`).

## Current Implementation
- I only hacked together the css & templates for fullwidth. The theme changes aren't that drastic and I will try to keep it backwards compatible.
- The css needs some rework, not that good at css.
- Subsections are added as a separate property on the doc_model object.
- Name and overview for a subsection (and subsubsection) is emitted before the tasks. See screenshots

## Questions
- Would the better option be adding the subsections to the children property instead of it's own? And then doing something like this in mustache?:
```
<li class="nav-group-name">
  <a class="nav-group-name-link" href="{{path_to_root}}{{url}}"><h{{level}}>{{section}}</h{{level}}></a>
  <ul class="nav-group-tasks">
    {{#children}}
      {{#children.count}}
  <ul class="nav-groups">
    {{#children}}
      {{> nav_section}}
    {{/children}}
  </ul>
      {{/children.count}}
      {{^children.count}}
    <li class="nav-group-task">
      <a class="nav-group-task-link" href="{{path_to_root}}{{url}}">{{name}}</a>
    </li>
    {{/children.count}}
  </ul>
</li>
```
- This however would then create complications with tasks, or would it?
- Should the full subsection be rendered out before the tasks come? IMO this would get quite lengthy with a lot of sub and subsubsections.
- Should the default theme autohide subsubsections (or even subsections)? i.e. make it show on clicking the parent section?
- Should there be no indentation in the navigation menu?
- Should I add some tests?

Generally, I would really appreciate some help with the css and generally the look & feel.

## Screenshots
![screen shot 2018-06-15 at 22 45 00](https://user-images.githubusercontent.com/5339762/41488888-3968bf96-70ee-11e8-8e63-9c1eb82a91d3.png)
![screen shot 2018-06-15 at 22 45 28](https://user-images.githubusercontent.com/5339762/41488890-399db6d8-70ee-11e8-86cd-8a3f17a63825.png)
![screen shot 2018-06-15 at 22 45 17](https://user-images.githubusercontent.com/5339762/41488889-39823584-70ee-11e8-8463-a65d994679d2.png)


Created with the following yaml:
```
module: TestFramework
abstract: Documentation/*.md
documentation: Guide.md
output: docs
theme: fullwidth
custom_categories:
  - name: Main
    children:
      - Test
      - name: Sub
        children:
          - WithEnum
          - name: SubSub
            children:
              - Guide
```

## Linked
Closes #624

johnfairh commented:
Thanks for the PR, brave place to start!  Will look at the code a bit more later.

I think the yaml format is OK; agree that just handling whatever depth the user wants is probably fine.

On the design, I'm not sure users will expect the child abstract hierarchy to appear in each parent abstract.  It feels to me as though it distracts from the discussion of the parent topic.  Maybe include the child subsections instead in a special Task at the bottom of the page task list that actually links in the normal way (with twisty) to the child section.
 
On the left nav, I guess I think yes the subsection structure should be reflected with some kind of indentation - wish I could help more on the style front!

You should use the misc_jazzy_features specs project to include tests for the feature.

galli-leo commented:
@johnfairh That‘s the integration tests right?

Yeah I am not too keen on each abstract as well. What‘s twisty? And yeah, the headers should probably link, didn‘t have any time yesterday to integrate that (though it should be easy).

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

1ec5 commented:
> What‘s twisty?

Pretty sure it’s a disclosure triangle, like this:

<details>
</details>

johnfairh commented:
(Sorry and ty @1ec5 for translating!)
Right, so basically include the subsection links as normal children in the body doc with a task heading (== the heading used for // MARK comments).

galli-leo commented:
@johnfairh Ah gotcha. Question would be, whether to keep the subsections seperate or add them to the tasks array.

johnfairh commented:
New thought: we should just include the subsection link in the children array where the user wrote it in their markdown, I think that's the least surprising behavior.  [Letting markdown specify headings is a separate issue!]

galli-leo commented:
@johnfairh I don't quite follow what you mean. Can you elaborate?

galli-leo commented:
@johnfairh Did you have some time to look at the implementation and decide whether to use a new property (like the current implementation) or use the existing children array? When that‘s decided, I can start working on tests and fixing up things.

Sorry if I am stressing you 😬, this would help a ton with our docs :)

johnfairh commented:
Sorry missed I owed you a response here!  Yes I looked through it, I think you've nicely captured the recursive building part of it and that it will work to treat nested categories as regular children.

PR #985 is a step on the way, deals with linking from category pages to items without an abstract; I'm planning to look at the left nav design/layout part next.

galli-leo commented:
@johnfairh  Cool, will try to get it changed to use children and tasks over the next few days.

JUSTINMKAUFMAN commented:
As mentioned elsewhere, I think this is a great idea. I’d love to see project directory parsing as a custom category source included in this PR. Wondering if I can help in any way?

johnfairh commented:
@galli-leo dw, I got it - take a look at [this branch](https://github.com/realm/jazzy/tree/jf-nested-dev) & see what you think.  I've treated the nested categories like everything else in the left nav so they are less heavy than in yours above, 🤷‍♂️ if this is OK - mostly been persuading the styles to nest reasonably.

@JUSTINMKAUFMAN would be a separate PR.  You would need to write a version of [this function](https://github.com/realm/jazzy/blob/83cebe59a40f5926662a5a6294ca1ba980149c5b/lib/jazzy/sourcekitten.rb#L63) that builds the category structure from the types returned by SourceKitten - they each have a `file` attribute with the pathname.

galli-leo commented:
@johnfairh Well also just got done, treating them like normal children 😬 . Also your mustache templates will never render, since you will have infinite recursion. If there is no children in the current context (i.e. a non category) mustache will try the parent context. Obviously that one has children, so it will render the parent nav again and again. See my branch for how I fixed that.

I like the idea of creating a new type, that makes a lot of sense.

I still think we should include the level, so that we can always use the correct h tag.

What do you think of my tasks implementation? When you click on a subcategories heading, it discloses that one's abstract now. We could also include the subsections in the normal tasks array, though I don't know how that will play out with mark comments. What do you think?

Live demo of the latest commit is here: https://galli.me/jazzy-demo/nested

johnfairh commented:
The templates are fine lol - easy to misread 😉 

Yep including the abstract if there is one (suspect often will not be) makes sense - should come out easily as part of treating the nested category like other children [I think this is what you mean by "including in the normal tasks array"]

The design you can see in eg. classes is to include just the abstract in the parent page and leave the next level of children to the next page down.

I misunderstood how much help you were asking for with the visual design here - will let you get on!

galli-leo commented:
@johnfairh Should I merge in your branch? Or are you planning on implementing this?

johnfairh commented:
Take as much/little as you need - if you need to do things a different way to get the visual effects you want for the headings and on the category pages then go ahead..

galli-leo commented:
@johnfairh So I integrated some of your changes (namely moving categories into their own class) into my pr. I think it's mostly ready for a review now, works quite well imo. Checkout https://galli.me/jazzy-demo/nested for a visual demo.

I made it, so that the Sub categories follow the same style as a mark heading and reveal the abstract as well as their items on clicking on it. Let me know, if I should also include the sub sub categories in that reveal.

Furthermore, should I try to fix those cyclomatic complexity errors from rubycop, or just ignore them with a comment? Not sure why it complains about the spacing for self.level = level as well, seems to be the same as your self.name = name.

galli-leo commented on lib/jazzy/source_category.rb:
```diff
@@ -0,0 +1,78 @@
+      self.abstract = Markdown.render(abstract)
+      self.children = group
+      self.parameters = []
+      self.mark       = SourceMark.new
```
Yeah, I left it from your code. Will remove.

galli-leo commented on lib/jazzy/source_category.rb:
```diff
@@ -0,0 +1,78 @@
+      self.children = group
+      self.parameters = []
+      self.mark       = SourceMark.new
+      self.level    = level
```
A gotcha, will fix.

galli-leo commented on lib/jazzy/sourcekitten.rb:
```diff
@@ -132,7 +81,9 @@ def self.make_doc_urls(docs)
-          doc.children = make_doc_urls(doc.children)
+          if doc.children.count > 0
+            doc.children = make_doc_urls(doc.children)
+          end
```
Might be leftover from your branch with the other changes, will remove.

galli-leo commented on lib/jazzy/sourcekitten.rb:
```diff
@@ -132,7 +81,9 @@ def self.make_doc_urls(docs)
-          doc.children = make_doc_urls(doc.children)
+          if doc.children.count > 0
+            doc.children = make_doc_urls(doc.children)
+          end
```
Would you rather have them moved to subdirectories or have a note in the config info?

I thought about adding a section to the README anyways.

johnfairh replied:
Require subcategory name uniqueness feels like a footgun/defect creator.  Putting the pages in subdirs should work.

galli-leo commented:
@johnfairh 
> Added some code nitpicks & a less trivial thought about category names.

Should have time tomorrow to fix those.

> I'm pretty sure that the left nav style needs to remain the same as now, certainly for sites that aren't using nested categories -- right now (no nested categories) it goes from:
> screen shot 2018-07-09 at 12 31 10 to screen shot 2018-07-09 at 12 32 32
> 
> From a design point of view, the sidebar is almost always less important than the body so needs not to have these big highly-spaced headings that are more eye-catching than stuff in the body.

Do you propose going back to just <a> tags or styling the headings differently? I agree it's a bit too big, I just wanted to make the distinction between subsections clear for demo purposes.

> The category headings in the body should probably use a different class or something from the mark headings: we know that some users do not like the mark headings and use custom CSS to suppress them.

Will add a custom CSS class. Should the style be still the same?

> Yes, strong default action is to fix the rubocop errors - they're usually pointing out good Ruby style points.

Should have time to fix those tomorrow as well.

> Either need apple + jony theme changes or docs + something to detect the incompatibility -- generated html looks OK at first but behaves pretty oddly.

Yeah I wanted to have one theme finalized, before working on the other ones. Will do those when fullwidth is good to go.

> I haven't checked how this looks in Dash.

I have taken a quick look and it seems fine, but the sections are unfortunately not nested inside dash. Not sure if that's possible or not.

johnfairh commented on lib/jazzy/source_category.rb:
```diff
@@ -0,0 +1,78 @@
+
+    def initialize(group, name, abstract, url_name, level = 1)
+      super()
+      self.type     = SourceDeclaration::Type.category
```
Ok.  `overview` is the type used for these category pages.  You could rename it (as long as this doesn't change existing URLs) but we don't need an additional type.

johnfairh commented:
> [sidebar]  just tags or styling the headings differently?

I guess find some other way of styling the 'sub'categories (same weight as the top categories?) - or leave them styled as they come out and rely on the indentation to indicate the hierarchy.

> [body category headings] Should the style be still the same?

Yes I think that makes sense.

> wanted to have one theme finalized, before working on the other ones. Will do those when fullwidth is good to go.

Sounds good!

galli-leo commented on lib/jazzy/source_category.rb:
```diff
@@ -0,0 +1,77 @@
+        children = category['children'].flat_map do |child|
+          if child.is_a?(Hash)
+            # Nested category, recurse
+            children, docs = group_custom_categories(docs, [child], level + 1)
```
I renamed docs_with_name to children to a) make rubocop happy and b) make it's purpose clearer.

Is that ok?

galli-leo commented:
@johnfairh I toned the navigation stylings down a bit. The only change now is font-weight, though I had to increase the font-size a little, or else the headings looked smaller (not sure why).

Before:
![screen shot 2018-07-12 at 21 51 03](https://user-images.githubusercontent.com/5339762/42658753-8b89839e-8626-11e8-8a84-871c5c080654.png)

After:
![screen shot 2018-07-12 at 22 02 16](https://user-images.githubusercontent.com/5339762/42658769-93227f3e-8626-11e8-975e-f1e153873491.png)

Can tone them down even more, if you think that's better. IMO having some differentiation through font-weight in addition makes it a bit easier to read.

johnfairh commented:
Code changes look great.  I think the `level` property is redundant now, can get it dynamically from the length of `documentation_path`.

I appreciate the work on the styling.  There's still additional vertical whitespace above heading elements (?); I would like the default weights to be as they are currently -- you can always override in your own projects.

galli-leo commented:
@johnfairh Sounds good, will apply the visual changes and level property over the next few days.

johnfairh commented:
Thanks!  This all looks good to me now -- code nits that you can get rid of the `attr_accessor :level` line, and the changelog is missing the double-space newline.

If you could update your demo site to reflect the most recent theme update + @jpsim can OK the design then I reckon good to update the other themes and merge it.

(sorry for the delay: this time my excuse is that GitHub doesn't email comment edits!)

galli-leo commented:
Fixed both and updated the demo site as requested.

BTW. maybe it would be a cool idea to host the generated integration specs somewhere? So a reviewer and even pr author can easily look at changes? Additionally, a visual diff would be even cooler. Is that something you have considered?

> (sorry for the delay: this time my excuse is that GitHub doesn't email comment edits!)

No worries! Out of curiosity, do you get notifications for commits to a pr? If not should I leave messages for new commits in the future?

galli-leo commented:
@johnfairh I will wait on the integration_specs pr until all themes are finished.

@jpsim If you are ok with the design, I can start on the other themes :)

galli-leo commented:
Bump

nrudnyk commented:
@galli-leo @johnfairh is there any progress/plans with this feature?

manand-ipratico commented:
Is there any update for this features?
thanks
